### PR TITLE
Fix: detecting when a db driver is missing

### LINKED ISSUE #124: detecting when a db driver is missing
In https://github.com/ploomber/jupysql/pull/120, we improved the error message displayed when failing to open a connection to a database. However, we're ignoring the error message content and printing a generic message. This leads to a confusing message when the connection string is valid, and the failure is due to another reason.

There are X reasons a connection might fail, but we should at least cover the most common ones. For example, missing a driver ([see this](https://github.com/ploomber/jupysql/pull/120#pullrequestreview-1289598402)). In such case, we should display a more appropriate message like "fix it by installing X package and try again"

I ran a few tests with some sample [URIs](https://docs.sqlalchemy.org/en/20/core/engines.html#postgresql) and the error is always the same `ModuleNotFoundError: No module named 'name'`.

The second case we must cover is when the plugin is missing. For example, if we run:

```
create_engine("duckdb://")
```

with duckdb-engine installed but duckdb *uninstalled*, we get: `ModuleNotFoundError: No module named 'duckdb'` (same as the use case described above).

However, if we uninstall duckdb-engine (regardless of duckdb installed or not), we get:

```
NoSuchModuleError: Can't load plugin: sqlalchemy.dialects:duckdb
```

`NoSuchModuleError` could be either a typo or an actual plugin missing (in this case, the duckdb-engine plugin is missing). So we should keep a list of plugins to distinguish between the two.

### PR DESCRIPTION
**Pull request by tonykploomber**
## Describe your changes

* Provide the suggestion message when the driver package is not found
  - Correct driver name, but not install in the user environment
  - Closely matching to correct driver name, but not quite correct
  - Neither the above cases and incorrect driver name

* Test cases to cover those 

### Case 1 - Missing driver

Correct driver name `pymysql`, no `pymysql` package installed

1. 
```
%sql mysql+pymysql://database_url
```

2.
```
UsageError: No module named 'pymysql'

To fix it, try to install package: pymysql

For technical support: https://ploomber.io/community
Documentation: https://jupysql.ploomber.io/en/latest/connecting.html
```

### Case 2 - Closely correct but incorrect driver name `pymysqll`

1. 
```
%sql mysql+pymysqll://
```

2.
```
UsageError: Can't load plugin: sqlalchemy.dialects:mysql.pymysqll

To fix it, perhaps you meant to use driver name: pymysql

For technical support: https://ploomber.io/community
Documentation: https://jupysql.ploomber.io/en/latest/connecting.html
```

### Case 3 - Neither close or correct driver name `something`

1. 
```
%sql mysql+something://
```

2.
```
UsageError: Can't load plugin: sqlalchemy.dialects:mysql.something

To fix it, make sure you are using correct driver name:
Ref: https://docs.sqlalchemy.org/en/20/core/engines.html#database-urls

For technical support: https://ploomber.io/community
Documentation: https://jupysql.ploomber.io/en/latest/connecting.html
```

## Issue ticket number and link
Closes #124

## Checklist before requesting a review

- [ ] I have performed a self-review of my code
- [ ] I have added thorough [tests](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#testing) (when necessary).
- [ ] I have added the right documentation in the [docstring](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#documenting-changes-and-new-features) and [changelog](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#changelog) (when needed)

----
:books: Documentation preview :books:: https://jupysql--127.org.readthedocs.build/en/127/

### CODE REVIEW

**Comment by tonykploomber** [comment]
Hi @edublancas 

Two things need to confirm with you:
1. Do you think I can still use the `UsageError` class to [present the error message](https://github.com/ploomber/jupysql/pull/127/files#diff-62741718c66525b0e94e4021ad561835184c9d1c00e2a12c00cb6c33346cabd4R170)?
I tried to use `ModuleNotFoundError` with our error message, however, it will show the trace of error execution in detail

**Comment by tonykploomber** [comment]
2. In current implementation, some existing test cases will fail

Those are in `src/tests/test_magic.py`

1. test_error_on_invalid_connection_string_format:
`%sql something://`
2. test_error_on_invalid_connection_string_with_possible_typo:
`%sql sqlite://`

Actually they are throwed by `sqlalchemy.create_engine`, since `sqlite` and `something` driver are not existing.

I think we should update the original test case, to except the user will see the `UsageError` with `To fix it, try to install the missing module` message, 

What do you think?

**Comment by tonykploomber** [comment]
Also please check the way how I maintain the [missing package dict](https://github.com/ploomber/jupysql/pull/127/files#diff-62741718c66525b0e94e4021ad561835184c9d1c00e2a12c00cb6c33346cabd4R14), if you think this approach is doable I can go ahead with other common drivers

**Comment by edublancas** [comment]
> Do you think I can still use the UsageError class to [present the error message](https://github.com/ploomber/jupysql/pull/127/files#diff-62741718c66525b0e94e4021ad561835184c9d1c00e2a12c00cb6c33346cabd4R170)?

Yeah, I think we can use UsageError, just do `raise UsageError(...) from e`, where e is the original exception. Python doesn't like it when you raise an exception while handling another exception

> Although the challenging part is: how do we know the driver namesomething is just the typo or there is a real missing driver not installed

we should keep a mapping of the most popular drivers (we can use the ones listed in sqlalchemy's documentation) and only display the "install package X" when there is a perfect match.

If there isn't a perfect match, we should let the other error show (the one I added here https://github.com/ploomber/jupysql/pull/120). Perhaps we can add to the generic message a "perhaps you meant to use X package?" and re-use the existing mapping. e.g.:

```
%sql dockdb://
```

```
Perhaps you meant duckdb?
```

to get similar matches, you can use: https://docs.python.org/3/library/difflib.html#difflib.get_close_matches

File: src/sql/connection.py [hunk 1]
```diff
@@ -3,8 +3,29 @@
...
 from sqlalchemy.engine import Engine
+from sqlalchemy.exc import NoSuchModuleError
 from IPython.core.error import UsageError
 
+PLOOMBER_SUPPORT_LINK_STR = (
+    "For technical support: https://ploomber.io/community"
+    "\nDocumentation: https://jupysql.ploomber.io/en/latest/connecting.html"
+)
+
+MISSING_PACKAGE_LIST_EXCEPT_MATCHERS = {
```
Comment (edublancas) [inline src/sql/connection.py:14]: yeah, this makes sense. 

why does duckdb contain `sqlalchemy.dialects` but the other doesn't? is this just how sqlalchemy displays the error?

**Review by edublancas — CHANGES_REQUESTED** [review]

**Comment by tonykploomber** [comment]
2/14 Update:

Progress:
- Add suggestion text for driver name is perfect matching -> imply driver is not installed, suggest user to install
- Add suggestion text for driver name is typo -> imply the typo, suggest user to type correct name

Block:
I was having a difficult time to simulate the package is not installed in unit-test phase.
For some driver, I faced some error when mocking some drivers weren't installed.

Tried to mock the missing package per:
- https://stackoverflow.com/a/60229056
- https://stackoverflow.com/questions/63865618/how-to-remove-a-library-with-monkeypatch-or-mock-in-pytest

None of those seems to work, our test cases will still be impacted by my system pip installed package.

I suggest moving this testing part into integration testing since we can manipulate the system environment on the container level

Will give the update soon

**Review by edublancas — CHANGES_REQUESTED** [review]
looks good, just one small change to make the error message clearer:

change:

```
UsageError: Can't load plugin: sqlalchemy.dialects:dockdb

To fix it, perhaps you meant to use driver name: duckdb

For technical support: https://ploomber.io/community
Documentation: https://jupysql.ploomber.io/en/latest/connecting.html
```

to:

```
UsageError: Can't load plugin: sqlalchemy.dialects:dockdb

Perhaps you meant to use the dialect: duckdb?

For technical support: https://ploomber.io/community
Documentation: https://jupysql.ploomber.io/en/latest/connecting.html
```

**Comment by tonykploomber** [comment]
error message clearer
fix in [930022a](https://github.com/ploomber/jupysql/pull/127/commits/930022ad93e85b12cbe8c49c83e8fde97e1c8ee6)

File: CHANGELOG.md [hunk 2]
```diff
@@ -2,6 +2,8 @@
 
 ## 0.5.6dev
 
+* [Feature] Shows missing driver package suggestion message ([#124](https://github.com/ploomber/jupysql/issues/124))
```
Comment (edublancas) [inline CHANGELOG.md:5]: FYI: there's no need to add the link manually. our deployment link takes care of it. just do (#NUMBER)

**Review by edublancas — APPROVED** [review]

**Pull request merged**